src/PowerShellEditorServices.Engine/Services/TextDocument/Handlers/CompletionHandler.cs

```diff
@@ -0,0 +1,320 @@
+        private static CompletionItem CreateCompletionItem(
+            CompletionDetails completionDetails,
+            BufferRange completionRange,
+            int sortIndex)
+        {
+            string detailString = null;
+            string documentationString = null;
+            string completionText = completionDetails.CompletionText;
+            InsertTextFormat insertTextFormat = InsertTextFormat.PlainText;
+
+            if ((completionDetails.CompletionType == CompletionType.Variable) ||
+                (completionDetails.CompletionType == CompletionType.ParameterName))
+            {
+                // Look for type encoded in the tooltip for parameters and variables.
+                // Display PowerShell type names in [] to be consistent with PowerShell syntax
+                // and now the debugger displays type names.
+                var matches = Regex.Matches(completionDetails.ToolTipText, @"^(\[.+\])");
+                if ((matches.Count > 0) && (matches[0].Groups.Count > 1))
+                {
+                    detailString = matches[0].Groups[1].Value;
+                }
+            }
+            else if ((completionDetails.CompletionType == CompletionType.Method) ||
+                     (completionDetails.CompletionType == CompletionType.Property))
+            {
+                // We have a raw signature for .NET members, heck let's display it.  It's
+                // better than nothing.
+                documentationString = completionDetails.ToolTipText;
+            }
+            else if (completionDetails.CompletionType == CompletionType.Command)
+            {
+                // For Commands, let's extract the resolved command or the path for an exe
+                // from the ToolTipText - if there is any ToolTipText.
+                if (completionDetails.ToolTipText != null)
+                {
+                    // Fix for #240 - notepad++.exe in tooltip text caused regex parser to throw.
+                    string escapedToolTipText = Regex.Escape(completionDetails.ToolTipText);
+
+                    // Don't display ToolTipText if it is the same as the ListItemText.
+                    // Reject command syntax ToolTipText - it's too much to display as a detailString.
+                    if (!completionDetails.ListItemText.Equals(
+                            completionDetails.ToolTipText,
+                            StringComparison.OrdinalIgnoreCase) &&
+                        !Regex.IsMatch(completionDetails.ToolTipText,
+                            @"^\s*" + escapedToolTipText + @"\s+\["))
+                    {
+                        detailString = completionDetails.ToolTipText;
+                    }
+                }
+            }
+            else if ((completionDetails.CompletionType == CompletionType.Folder) &&
+                     (completionText.EndsWith("\"", StringComparison.OrdinalIgnoreCase) || completionText.EndsWith("'", StringComparison.OrdinalIgnoreCase)))
```

Comment:
Not sure if `EndsWith` optimizes this, but maybe better to have a static method like:

```csharp
bool EndsWithQuote(string text)
{
    if (string.IsNullOrEmpty(text))
    {
        return false;
    }

    char lastChar = text[text.Length - 1];
    return lastChar == '"' || lastChar == '\'';
}
```

Reply:
added